lib/spack/spack/cmd/view.py

```diff
@@ -0,0 +1,361 @@
+            help='Add package files to a filesystem view via symbolic links.'),
+        ssp.add_parser(
+            'hardlink', aliases=['hard'],
+            help='Add packages files to a filesystem via via hard links.'),
+        ssp.add_parser(
+            'remove', aliases=['rm'],
+            help='Remove packages from a filesystem view.'),
+        ssp.add_parser(
+            'statlink', aliases=['status', 'check'],
+            help='Check status of packages in a filesystem view.')
+    ]
+    # All these options and arguments are common to every action.
+    for act in file_system_view_actions:
+        act.add_argument('path', nargs=1,
+                         help="Path to file system view directory.")
+        act.add_argument('specs', **specs_opts)
+
+    # The formatted print action.
+    act = ssp.add_parser('print',
+                         help="Print a string to stdout based on given format")
+    act.add_argument('format', nargs=1,
+                     help="Format describing per-package printout.")
+    act.add_argument('specs', **specs_opts)
+
+    return
+
+
+def assuredir(path):
+    'Assure path exists as a directory'
+    if not os.path.exists(path):
+        os.makedirs(path)
+
+
+def relative_to(prefix, path):
+    'Return end of `path` relative to `prefix`'
+    assert 0 == path.find(prefix)
+    reldir = path[len(prefix):]
+    if reldir.startswith('/'):
+        reldir = reldir[1:]
+    return reldir
+
+
+def transform_path(spec, path, prefix=None):
+    'Return the a relative path corresponding to given path spec.prefix'
+    if os.path.isabs(path):
+        path = relative_to(spec.prefix, path)
+    subdirs = path.split(os.path.sep)
+    if subdirs[0] == '.spack':
+        lst = ['.spack', spec.name] + subdirs[1:]
+        path = os.path.join(*lst)
+    if prefix:
+        path = os.path.join(prefix, path)
+    return path
+
+
+def purge_empty_directories(path):
+    '''Ascend up from the leaves accessible from `path`
+    and remove empty directories.'''
+    for dirpath, subdirs, files in os.walk(path, topdown=False):
+        for sd in subdirs:
+            sdp = os.path.join(dirpath, sd)
+            try:
+                os.rmdir(sdp)
+            except OSError:
+                pass
+
+
+def filter_exclude(specs, exclude):
+    'Filter specs given sequence of exclude regex'
+    to_exclude = [re.compile(e) for e in exclude]
+
+    def exclude(spec):
+        for e in to_exclude:
+            if e.match(spec.name):
+                return True
+        return False
+    return [s for s in specs if not exclude(s)]
+
+
+def flatten(seeds, descend=True):
+    'Normalize and flattend seed specs and descend hiearchy'
+    flat = set()
+    for spec in seeds:
+        if not descend:
+            flat.add(spec)
+            continue
+        flat.update(spec.normalized().traverse())
+    return flat
+
+
+def spec2dict(spec):
+    'Convert info in a spec into a simple dictionary.'
```

Comment:
You raise a few issues, some I've been pondering as well.

For UI, might it be better to move `spack view print` to a first-level sub-command: `spack print` or `spack query` (to name it after `dpkg-query` from which I stole this concept).  I think "yes" but this means moving the visitor pattern out of `spack.cmd.view` and into, say, `spack.visitor` so it can be shared.  This is probably a good thing in any case as `spack.cmd`s should be merely adapters between UI and core Spack functionality.

If `Spec.format` could be extended to add `spec.package` parts and the special-case munging (like turning `spec.dependencies.keys()` into a comma separated list, similar for `spec.rpath` then I think `spec2dict()` can/should happily go away.

The last "belonging" issue you raise goes back to something Tom's original suggestion also digs at.  I think I see this visitor pattern differently than you two do.  To me the visitor is a parameter of the visitation.  Specifically, it's a mapping from a parameter given by the user as a string to a function defined in some Python code, but it's a parameter nonetheless.  Seen as such it makes more sense to me to have a generic `view` (or better `visit`) sub-command which accepts a zoo of visitor names to apply as visitor functions.  But, I do agree that from a UI perspective, deep sub-sub-commands can become annoying. 

So, as long as this PR is getting, I suggest going a little further:

- extend `Spec.format()` to be some sensible union of what it is now and what `spec2dict` is.
- move the visitor pattern into `spec.visitor` 
- move `spack view print` into `spack query`

Maybe I can start by doing just that last one as it affects UI.  Then we can merge this PR and I can work on the other two in a follow up.